lib/topology.ml

```diff
@@ -0,0 +1,250 @@
+
+  let union a b =
+    v
+      ~affinity:(CPUSet.union a.affinity b.affinity)
+      ~memfree:(Int64.add a.memfree b.memfree)
+
+  let min_memory r1 r2 = {r1 with memfree= min r1.memfree r2.memfree}
+
+  let pp_dump =
+    Fmt.(
+      Dump.record
+        [ Dump.field "affinity" (fun t -> t.affinity) CPUSet.pp_dump
+        ; Dump.field "memfree" (fun t -> t.memfree) int64 ])
+end
+
+module NUMARequest = struct
+  type t = {memory: int64; vcpus: int}
+
+  let v ~memory ~vcpus =
+    if Int64.compare memory 0L < 0 then
+      invalid_arg
+        (Printf.sprintf "NUMARequest: memory must be > 0: %Ld" memory) ;
+    if vcpus < 0 then
+      invalid_arg (Printf.sprintf "vcpus cannot be negative: %d" vcpus) ;
+    {memory; vcpus}
+
+  let fits requested available =
+    Int64.compare requested.memory available.NUMAResource.memfree <= 0
+    && CPUSet.(cardinal available.NUMAResource.affinity >= requested.vcpus)
+
+  let shrink a b =
+    v
+      ~memory:(max 0L (Int64.sub a.memory b.NUMAResource.memfree))
+      ~vcpus:(max 0 (a.vcpus - CPUSet.cardinal b.NUMAResource.affinity))
+
+  let pp_dump =
+    Fmt.(
+      Dump.record
+        [ Dump.field "memory" (fun t -> t.memory) int64
+        ; Dump.field "vcpus" (fun t -> t.vcpus) int ])
+end
+
+(** [seq_range a b] is the sequence of numbers between [a, b). *)
+let seq_range a b =
+  let rec next i () = if i = b then Seq.Nil else Seq.Cons (i, next (i + 1)) in
+  next a
+
+(** [gen_2n n] Generates all non-empty subsets of the set of [n] nodes. *)
+let seq_gen_2n n =
+  (* A node can either be present in the output or not,
+   * so use a loop [1, 2^n) and have the [i]th bit determine that.
+   * *)
+  let of_mask i =
+    seq_range 0 n |> Seq.filter (fun bit -> (i lsr bit) land 1 = 1)
+  in
+  seq_range 1 (1 lsl n) |> Seq.map of_mask
+
+(** [seq_sort ~cmp s] sorts [s] in a temporary place using [cmp]. *)
+let seq_sort ~cmp s =
+  let a = Array.of_seq s in
+  Array.fast_sort cmp a ; Array.to_seq a
+
+(** [seq_append a b] is the sequence [a] followed by [b] *)
+let seq_append (a : 'a Seq.t) (b : 'a Seq.t) =
+  let rec next v () =
+    match v () with
+    | Seq.Nil ->
+        b ()
+    | Seq.Cons (x, xs) ->
+        Seq.Cons (x, next xs)
+  in
+  next a
+
+module NUMA = struct
+  type node = Node of int
+
+  (* no mutation is exposed in the interface,
+   * therefore this is immutable *)
+  type t =
+    { distances: int array array
+    ; cpu_to_node: node array
+    ; node_cpus: CPUSet.t array
+    ; all: CPUSet.t
+    ; candidates: (float * node Seq.t) Seq.t }
+
+  let node_of_int i = Node i
+
+  let node_distances d nodes =
+    let dists =
+      nodes
+      |> Seq.flat_map (fun n1 -> nodes |> Seq.map (fun n2 -> d.(n1).(n2)))
+    in
+    let count, max_dist =
+      Seq.fold_left
+        (fun (count, maxv) e -> (count + 1, max maxv e))
+        (0, min_int) dists
+    in
+    let avg_dist = (Seq.fold_left ( + ) 0 dists |> float) /. float count in
```

Comment:
i assume you are using `Seq` because there are a lot of nodes?

if that is the case, then it would make sense to do the count, max_dist and avg_dist in one fold

Reply:
I was using Seq to make the choice of data structure abstact, if we want to move to a different one in the future it can still expose a Seq. You are right that this could be done in one go.